jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp

```diff
@@ -0,0 +1,76 @@
+/// Author: Diffblue Ltd.
+
+#include "generic_parameter_specialization_map.h"
+
+std::size_t generic_parameter_specialization_mapt::insert(
+  const std::vector<java_generic_parametert> &parameters,
+  std::vector<reference_typet> types)
+{
+  PRECONDITION(!parameters.empty());
+  std::stringstream msg;
+  msg << "mapping <";
+  join_strings(
+    msg,
+    parameters.begin(),
+    parameters.end(),
+    ", ",
+    [](const java_generic_parametert &param) { return param.get_name(); });
+  msg << ">";
+  std::string msg_str = msg.str();
+  const auto first_param_it =
+    param_to_container.find(parameters.front().get_name());
+  std::size_t container_index;
+  if(first_param_it == param_to_container.end())
+  {
+    container_index = container_to_specializations.size();
+    container_to_specializations.emplace_back();
+    std::size_t param_index = 0;
+    for(const java_generic_parametert &parameter : parameters)
+    {
+      const auto result = param_to_container.emplace(
+        parameter.get_name(), container_paramt{container_index, param_index++});
+      INVARIANT(
+        result.second, "Some but not all type parameters already mapped");
+    }
+  }
+  else
+  {
+    container_index = first_param_it->second.container_index;
+    std::size_t param_index = 0;
+    for(const java_generic_parametert &parameter : parameters)
+    {
+      const auto param_it = param_to_container.find(parameter.get_name());
+      INVARIANT(
+        param_it != param_to_container.end(),
+        "Some but not all type parameters already mapped");
+      INVARIANT(
+        param_it->second.container_index == container_index,
+        "Not all type parameters assigned to same container");
+      INVARIANT(
+        param_it->second.param_index == param_index++,
```

Comment:
Eeep! Let's not have `INVARIANT` conditions with side-effects, there are some compilation settings where the first arg is discarded